### ISSUE TITLE
Parsing speed

### ISSUE THREAD

**Issue by sbromberger** [andrewcooke/ParserCombinator.jl#5]
Currently, parsing http://www-personal.umich.edu/~mejn/netdata/celegansneural.zip is taking a long time (~ 20 minutes) on my machine. This is a 15k-line file with 302 nodes and 2359 edges.

**Comment by andrewcooke** [comment]
ok, i'll have a look.  that sounds crazy slow, so hopefully there's some fix.

i'll also look at the overflow thing.

cheers,
andrew

**Comment by andrewcooke** [comment]
hmmm.  i'm getting 13s on my machine.  this may be because the vcat thing was so inefficient.

this is the code i used:

```julia
using ParserCombinator.Parsers.GML
io = open("gml/celegansneural.gml")
@time x = parse_dict(io; debug=false)
@test length(x) == 2
@test x[:Creator] == "Mark Newman on Thu Aug 31 12:59:09 2006"
@test length(x[:graph][1]) == 3
@test length(x[:graph][1][:node]) == 297
@test length(x[:graph][1][:edge]) == 2359
```

and the output is

```
  12.903 seconds      (25727 k allocations: 1155 MB, 11.93% gc time)
```

note that the above parses the file directly (takes an IO instance, rather than a string).  that's new, and error messages aren't going to work correctly.  i'll fix that this weekend.  but i don't think it should change the speed.

**Comment by andrewcooke** [comment]
i just noticed the number of nodes is slightly different to what you say.  297 checks with grep/wc, so i think i am ok.  is it possible you have a file that is corrupt in some way?  if the file has an error and the parser is backtracking trying to find a way to parse it correctly that could take a long time.  i've tried to make the parser fail on errors, but backtracking parsers are not great in this case.  if you do have a broken file, could you send me it, and i'll see if i can improve behaviour in that case?

**Comment by andrewcooke** [comment]
btw the other file (polblogs or whatever it's called) takes 120s for me.

**Comment by andrewcooke** [comment]
heh.  changing the caching strategy gets those down to 6s and 46s. now in git.

**Comment by andrewcooke** [comment]
(sorry; i may not have pushed all my changes to github - as of this post, however, github is current)

**Comment by sbromberger** [comment]
I just pulled the latest master and am still seeing pretty bad performance on this graph: https://gist.github.com/sbromberger/8c8edfc808a7f2f091a9

It's still reading after ~10 minutes.

```
    function readgml(filename::String)
        f = open(readall,filename)
        p = parse_dict(f)
        g1 = p[:graph][1]
        dir = Bool(get(g1, :directed, 0))

        nodes = [x[:id] for x in g1[:node]]
        mapping = Dict{Int,Int}()
        for (i,n) in enumerate(nodes)
            mapping[n] = i
        end
        if dir
            g = DiGraph(length(nodes))
        else
            g = Graph(length(nodes))
        end
        sds = [(Int(x[:source]), Int(x[:target])) for x in g1[:edge]]
        for (s,d) in (sds)
            add_edge!(g, mapping[s], mapping[d])
        end
        return g
    end
```

**Comment by andrewcooke** [comment]
ok, slow down a bit.  what about the ones you were having problems with previously?  are they now ok?  what about the discrepancy in the number of nodes?

**Comment by sbromberger** [comment]
Sorry - I created my own gml file since the other ones were multigraphs and failed during the add_edge! stage. Let's use this one as a reference point if that's ok with you.

**Comment by andrewcooke** [comment]
ok the following code, just doing the parsing, and doing an initial pass for the JIT
```julia
println("10k-49963")
io = open("gml/10k-49963.gml")
parse_raw(io; debug=false)
io = open("gml/10k-49963.gml")
@time parse_raw(io; debug=false)
```
took 230s on my laptop.  the latest in git gets that down to 80s.  almost a factor of 3 faster.  but i don't think i can improve much on that.  if you want faster, you'll need to write something by hand or use a more "heavyweight" parser (like a julia wrapper round antlr). (parse_raw or parse_dict makes very little difference).

part of the problem is that the code is untyped.  i'll try think how to change that, but i doubt it's really possible.  there might be a way to add type annotations in various places in a grammar, but i suspect it's going to be hard.  part of the problem is that the file format is more flexible than you think looking at your example - i could likely write a faster parse that parsed that file, but not all GML files.

**Comment by andrewcooke** [comment]
i had some more ideas overnight i will try today, but i'm also concerned about the differences we see in timings.  this last file took 4m on my machine and over 20 on yours.  either my machine is wonderful, yours is sucky, or your additional code is expensive.  the code looks ok to me, but i'd prefer you looked at timing data for just the parser when reporting issues with the parser.

**Comment by sbromberger** [comment]
> either my machine is wonderful, yours is sucky, or your additional code is expensive

Could be. My machine is
```
Julia Version 0.4.0-dev+6112
Commit 189a043 (2015-07-20 17:52 UTC)
Platform Info:
  System: Darwin (x86_64-apple-darwin14.4.0)
  CPU: Intel(R) Core(TM) i7-5557U CPU @ 3.10GHz
  WORD_SIZE: 64
  BLAS: libopenblas (USE64BITINT DYNAMIC_ARCH NO_AFFINITY Haswell)
  LAPACK: libopenblas
  LIBM: libopenlibm
  LLVM: libLLVM-3.3
```

I'm currently trying with just the parser to eliminate any delay in my LightGraphs code. Will report back shortly.

**Comment by andrewcooke** [comment]
i don't think that's worse than mine (i have no idea about model numbers, but they're both i7's and yours has a larger "number" than mine)

```
Julia Version 0.4.0-dev+6026
Commit cf6b3c3 (2015-07-17 17:32 UTC)
Platform Info:
  System: Linux (x86_64-suse-linux)
  CPU: Intel(R) Core(TM) i7-4600U CPU @ 2.10GHz
  WORD_SIZE: 64
  BLAS: libopenblas (USE64BITINT DYNAMIC_ARCH NO_AFFINITY Haswell)
  LAPACK: libopenblas
  LIBM: libopenlibm
  LLVM: libLLVM-3.3
``

**Comment by sbromberger** [comment]
Getting

```
julia> @time z=parse_raw(io; debug=false);
  87.422 seconds      (235 M allocations: 9603 MB, 5.26% gc time)
```
with the latest (pulled ~20 mins ago) ParserCombinator.

**Comment by andrewcooke** [comment]
oh good.  so that's comparable to what i see.  i've had some more ideas, as i said, but am struggling a bit to get it to work.

**Comment by sbromberger** [comment]
Cool. Not to switch gears, but if you wanted to try parsing GraphML (different from GML), it might be a good second project. We've got a GraphML parser already but it depends on LightXML and I'd love to be able to reduce dependencies.

**Comment by andrewcooke** [comment]
yeah, i actually had already considered it, but parsing XML well is quite hard and likely other packages do it better than i could. :o(

**Comment by sbromberger** [comment]
Ah, ok. No worries. Let's focus on getting GML performant. I'm currently profiling to see where allocations are occurring.

**Comment by andrewcooke** [comment]
if there's a fix / work-around to [this](https://groups.google.com/forum/#!topic/julia-users/GK9hMQ8ymY4) i'm pretty sure i have at least another factor of 2.

**Comment by andrewcooke** [comment]
ok, have an issue now - https://github.com/JuliaLang/julia/issues/12262#issuecomment-123550452 - and i am rewriting the parser so that i can special case regex with strings as efficiently as possible.  this will probably take a few days.

**Comment by andrewcooke** [comment]
latest in git is now at 70s for the 10k file (was 80s) - that's mainly from being cleaner about types in various places.  when the patch comes through (i am not 100% sure it will, but it looks hopeful) the code is now in a state where i can take advantage of that.

(note largely to myself - currently, passing a large string to regex involves copying data, so is expensive.  to avoid that we're using line by line parsing.  when the patch comes through we can parse the entire string, so can write regexps that cross line boundaries.  that will let us match and discard following spaces, instead of discarding space as a separate matcher.  which should reduce work by around 50%).

**Comment by andrewcooke** [comment]
yep!  with that patch (and some minor cleanup) the time has halved.  i can now add some more fixes that should reduce time further.  hopefully this will get into 4.0.0.

**Comment by andrewcooke** [comment]
OK, finally...  The patch referenced above works great.  It will hopefully be released in v0.4.0 and the code switches on that version.  To test that code (eg with malmaud's branch - using trunk will result in memory overflow) set "FAST_REGEX=true" in the Main module (ie top of first file) before loading ParserCombinator.

Here are the current timing results on my laptop with the 10k graph from the gist linked above (first for trunk, and then with FAST_REGEX enabled in @malmaud's branch):

```
andrew@laptop:~/project/ParserCombinator/test> julia-trunk runtests.jl 
10k-49963
  41.203 seconds      (115 M allocations: 4667 MB, 5.65% gc time)
andrew@laptop:~/project/ParserCombinator/test> ~/pkg/julia-malmaud/julia runtests.jl 
10k-49963
  16.641 seconds      (45906 k allocations: 1456 MB, 3.86% gc time)
```

(The first timing for that data in the text above, on my laptop, was 230s - we've come a long way!)

There's still room for improvement.  The two main sources of allocations now are (1) related to my stack use in the trampoline and (2) something to do with calling regex, perhaps the issue described by malmaud in the linked thread.  I have some ideas to reduce (1), but it will have to wait, and to be honest I'd prefer that the code in malmaud's branch is merged rather than tackling (2) right now.  I think 0.4.0 release is soon...

This is in git now.

**Comment by andrewcooke** [comment]
So closing this! :o)

**Comment by sbromberger** [comment]
Thanks for ALL your work tracking this down. It was certainly way more complicated than I ever imagined. I'll git pull and post times.

**Comment by andrewcooke** [comment]
No problem - it was a lot of fun and has improved the code hugely.  Unfortunately I work next week, so this is the end of my free time for now.  I hope your results are positive!

**Comment by sbromberger** [comment]
Any idea when 0.4 will be patched?

**Comment by andrewcooke** [comment]
No, I'll ask in that thread.

**Comment by sbromberger** [comment]
Haven't gotten the 0.4 patches in yet, but getting a nice speedup without:

```
julia> @time g = LightGraphs.readgml(f)
  44.216 seconds      (118 M allocations: 4787 MB, 4.38% gc time)
{10000, 49963} directed graph
```
Can't wait to see what happens with the patch!

**Comment by sbromberger** [comment]
Well, it looks like the patch was committed 20 hours ago, so I rebuilt, but I'm not seeing the performance improvement:

```
julia> @time g = LightGraphs.readgml(f)
  48.021 seconds      (118 M allocations: 4790 MB, 4.58% gc time)
{10000, 49963} directed graph
```

**Comment by andrewcooke** [comment]
see comment on FAST_REGEX above.

**Comment by andrewcooke** [comment]
also, it will be fastest given a string.  it reads files line by line so can't use a regex over multiple lines.  i should change that...

**Comment by sbromberger** [comment]
Oh. Duh - I missed the `FAST_REGEX` bit. Sorry! 

```
julia> @time g = LightGraphs.readgml(f)
  12.363 seconds      (31634 k allocations: 1040 MB, 3.98% gc time)
{10000, 49963} directed graph
```

:+1:

**Comment by sbromberger** [comment]
Is there a disadvantage to setting `FAST_REGEX=true` without the corresponding patch, or should I wrap it in a conditional version check?

**Comment by sbromberger** [comment]
(also, I can't figure out why this isn't working. I've tried `const FAST_REGEX=true` and `global FAST_REGEX=true` in the LightGraphs.jl main file (the module) but it's not working. That is, LightGraphs.FAST_REGEX prints true but ParserCombinator is reverting to slow behavior.

**Comment by andrewcooke** [comment]
when 0.4.0 is released (which is soon, afaik), you won't need FAST_REGEX.

as far as i can see, that's the best that we can do.  if you hard-code it then anyone who's not using the patched julia will get a memory leak and the program will crash.

if you want to override the default behaviour then you must set FAST_REGEX in the Main module.  not in LightGraphs.  i realise that's not much use to you, but given the above, it was intended only for testing before the 0.4.0 release.

**Comment by sbromberger** [comment]
I've made a PR to do the FAST_REGEX test use `v"0.4.0-dev+6325"` instead of `v"0.4.0"` - this shouldn't have any unintended side effects. However, I note that Travis is failing (unrelated to this PR) on 0.3.

**Comment by sbromberger** [comment]
See https://github.com/andrewcooke/ParserCombinator.jl/pull/8

**Comment by andrewcooke** [comment]
ah, i had no idea that was possible.  thanks.  i'll merge it and fix 0.3 this evening.

**Comment by sbromberger** [comment]
Thanks! Also, please tag a new version number so I can REQUIRE the correct version in LightGraphs :)

**Comment by andrewcooke** [comment]
1.6.2 has your patch + the fix for 0.3
